### PR TITLE
perf(staker): encode liquidity history as uint128

### PR DESCRIPTION
**Pull request by junghoon-vans**
## Summary
- Store staked-liquidity history checkpoints as fixed-width 16-byte big-endian uint128 strings.
- Preserve predecessor lookup, same-timestamp overwrite, zero transitions, and caller/result isolation.

## Decision
- Proceed with this fresh-deployment-only contract change: the measured storage cash-flow improvement for retained liquidity history exceeds the measured execution-gas premium in the 20-cycle scope.
- This recommendation is not deployment approval. Scope, compatibility, correctness, and deployment costs remain separate gates.

## Context
- Position and pool liquidity are constrained to the uint128 range; storing each history value as uint256 carries unnecessary serialized overhead.
- Outside-accumulation compression is intentionally excluded; it is provided by the parent PR.

## Changes
- Add bounded uint128 encode/decode helpers alongside the parent PR's uint256 codec.
- Encode staked-liquidity checkpoints on write and decode them at the resolver boundary.
- Reject nil, values above uint128, and malformed encoded values.
- Add boundary, byte-order, overflow, lookup, overwrite, zero-transition, and mutation-isolation coverage.

## Performance Evidence
Measured in [Tracker #51](https://github.com/gnoswap-labs/gnoswap-performance-tracker/pull/51). Delta is candidate minus baseline: negative storage delta is an improvement; positive gas delta is additional execution gas. Storage is the tracker's net per-realm storage diff during each workload, not peak storage footprint or total writes. CPU cycles are GnoVM counters, not wall-clock time.

| Workload | Baseline gas | Candidate gas | Gas Δ | Gas Δ % | Storage Δ (B) | CPU-cycle Δ |
|---|---:|---:|---:|---:|---:|---:|
| 20 unstake/restake cycles (40 calls) | 850,274,445 | 859,158,520 | +8,884,075 | +1.04% | -33,363 | +8,773,720 |
| Initial stake | 14,166,770 | 14,321,808 | +155,038 | +1.09% | -831 | +153,952 |
| History append 64→65 | 26,185,538 | 26,447,937 | +262,399 | +1.00% | -837 | +284,734 |
| History append 65→66 | 17,175,999 | 17,331,016 | +155,017 | +0.90% | -837 | +153,952 |
| CollectReward, 1 external token, later | 21,125,801 | 21,394,217 | +268,416 | +1.27% | 0 | +261,564 |
| CollectReward, 4 external tokens, later | 42,769,167 | 43,305,999 | +536,832 | +1.26% | 0 | +523,128 |
| Same-timestamp stake + unstake | 22,079,030 | 22,433,407 | +354,377 | +1.61% | 0 | +351,498 |
| Swap, 50 staked crossings, halving | 445,343,810 | 445,498,827 | +155,017 | +0.03% | -837 | +153,952 |
| Swap, 10 staked crossings, no halving | 94,229,033 | 94,339,116 | +110,083 | +0.12% | 0 | +110,358 |
| Two-hop exact-out control | 73,105,182 | 73,105,182 | 0 | +0.00% | 0 | 0 |

For the 40-call cycle group, Staker-only net growth falls from 75,954 B to 42,591 B (43.93% less); all-realm net growth falls from 56,759 B to 23,396 B. Workloads overlap, so rows must not be summed into a protocol-wide estimate.

### Priced 20-cycle result

| Component | Candidate − baseline |
|---|---:|
| Storage deposit/refund cash flow | -3.336300 GNOT |
| Execution gas equivalent | +0.008884 GNOT |
| Net estimated gain | 3.327416 GNOT |

The measured later `CollectReward` premium is +0.000268 GNOT for one external token and +0.000537 GNOT for four. The cycle result is therefore not a lifetime saving that ignores recurring read costs.

## Reproduction
- Baseline contract: `b12f0f8760a3e88e9539b0c8cfd32b38a5297c68`
- Measured candidate: `0ac0d4482191418262c03f369b8334a9f8f26b37`
- Tracker revision: `54aa519292adc7656601dd447f346115369fe4d1`
- Pinned metric Gno: `bac78a97b20e12ba471bdd73a41af31bd94bdc12`

```sh
make init
make compare-metric-force 0ac0d4482191418262c03f369b8334a9f8f26b37 b12f0f8760a3e88e9539b0c8cfd32b38a5297c68
python3 scripts/price_liquidity_metrics.py b12f0f8760a3e88e9539b0c8cfd32b38a5297c68 0ac0d4482191418262c03f369b8334a9f8f26b37 reports/metric/compares/pricing_0ac0d44_b12f0f8.json
```

## Verification
- Tracker #51 completed the paired comparison across 199 matching labels and validated all 597 deltas and applicable percentages.
- Tracker #51 ran the complete `gno.land/p/gnoswap/utils` and `gno.land/r/gnoswap/staker/v1` package suites with pinned Gno.
- This stacked branch passes `make test PKG=gno.land/p/gnoswap/utils`.

## Compatibility
- Parent PR: https://github.com/gnoswap-labs/gnoswap/pull/1446
- This PR must merge after #1446 because it builds on the shared binary codec changes.

## Deployment Notes
- Fresh deployments only. Existing staked-liquidity histories are not decoded or migrated.
- The tracker measurement excludes migration and deployment cost. Do not deploy this reader on legacy pointer-based history or roll back to a pointer-only reader after encoded history is written.
- Event payloads and public event schemas are unchanged; no batch, indexer, API, or frontend companion change is required.

## Cost Assumptions
- Pearl observation on 2026-09-09: 1,000 gas = 0.000001 GNOT and 1 byte = 0.000100 GNOT.
- Storage pricing assumes fresh-state 100 ugnot/B deposit/refund ratios and includes both deposit and refund effects.
- Values are same-rate repricing estimates, not wallet debits. With the same valid fixed `GasFee`, the gas cash delta is zero.
- Transaction overhead, deployment/migration cost, and capital time value are excluded.

## Summary by CodeRabbit

* **New Features**
  * Added support for encoding and decoding unsigned 128-bit values in a fixed-width format.
  * Staked liquidity history now stores and retrieves checkpoint values using the standardized 128-bit format.
  * Added validation for malformed values and numbers exceeding the supported range.

* **Tests**
  * Expanded coverage for encoding, decoding, checkpoint history behavior, boundary values, and invalid inputs.

### CODE REVIEW

File: contract/p/gnoswap/utils/codec_test.gno [hunk 1]
```diff
@@ -186,3 +186,68 @@ func TestDecodeUint256RejectsInvalidLength(cur realm, t *testing.T) {
...
+				uassert.Equal(t, tt.want[i], encoded[i])
+			}
+
+			decoded := DecodeUint128(string(tt.want[:]))
+			uassert.Equal(t, tt.value.ToString(), decoded.ToString())
+		})
+	}
+}
+
+func TestUint128CodecRejectsInvalidInput(cur realm, t *testing.T) {
```
Comment (notJoon) [inline contract/p/gnoswap/utils/codec_test.gno:280]: this `cur realm` param seems unnecessary 
```suggestion
func TestUint128CodecRejectsInvalidInput(t *testing.T) {
```

**Review by notJoon — APPROVED** [review]

**Review by jinoosss — COMMENTED** [review]
The intent and direction of the modification are clear, so it would be beneficial to review it once more.
However, while storage savings can be realized once per block, there is a potential disadvantage regarding the gas costs for multiple swaps within the same block.
